001-collection-backend-dispatching.md

```diff
@@ -0,0 +1,165 @@
+    ddf = dask.dataframe.read_parquet("./tmpdir")
+
+with dask.config.set({"array.backend": "cupy"}):
+    # Produce a cupy-backed collection
+    darr = dask.array.ones(10, chunks=(5,))
+```
+
+Since it is unlikely that an alternative backend will support all numpy- or pandas-based data-creation functions available in the collection API, we also propose a "backend-options" field containing "allow-fallback" and "warn-fallback" subfields. When the "allow-fallback" field is set to `True` (default), then the backend's designated fallback class will be used to perform IO, and the result will be moved from the fallback backend. The user should also have the option to enable or disable warnings when this fallback behavior occurs:
+
+
+```python
+options = {
+    # Allow user to specify if a fallback backend library
+    # should be used, and if a warning should be raised when
+    # this occurs:
+    "allow-fallback" : True,
+    "warn-fallback" : True,
+}
+
+dask.config.set({"dataframe.backend-options": options})
+```
+
+The specific configuration options proposed here are certainly not set in stone.  However, we do feel that the user should have complete control over fallback behavior.  For example, the user should be able to specify if falling back to "numpy"/"pandas" should result in a warning or error message, or if it should be ignored altogether.
+
+
+### Registering a New Backend (`DaskBackendEntrypoint`)
+
+In order to allow backend registration outside of the Dask source code, we propose that Dask approximately follow [the approach taken by xarray for custom backend interfaces](https://xarray.pydata.org/en/stable/internals/how-to-add-new-backend.html). That is, external libraries should be able to leverage "entrypoints" to tell Dask to register compatible backends in Dask-Array and Dask-DataFrame at run time. To this end, the external library could be expected to define all dispatch IO logic within a `DaskBackendEntrypoint` subclass.  For example, a cudf-based subclass would look something like the `CudfBackendEntrypoint` definition below:
+
+```python
+class CudfBackendEntrypoint(DaskBackendEntrypoint):
+
+    @cached_property
+    def fallback(self):
+        """Fallback entrypoint object to use for missing attributes
+
+        Returning anything other than ``None`` requires that
+        ``move_from_fallback`` be properly defined.
+        """
+        return PandasBackendEntrypoint()
+
+    def move_from_fallback(self, ddf):
+        """Move a Dask collection from the fallback backend"""
+        if isinstance(ddf._meta, pd.DataFrame):
+            return ddf.map_partitions(cudf.DataFrame.from_pandas)
+        elif isinstance(ddf._meta, pd.Series):
+            return ddf.map_partitions(cudf.Series.from_pandas)
+        return ddf
+    ...
+
+    def read_json(self, *args, engine=None, **kwargs):
+        return self.fallback.read_json(*args, engine=cudf.read_json, **kwargs)
+
+    def read_orc(self, *args, **kwargs):
```

Comment:
Ok that seems good to me. It just means that there might be a lot of duplication right? But I guess there is probably already a lot of duplication because dask_cudf already exists.